File: test/test.01.cpp
// Created by Grigore Stefan <g_stefan@yahoo.com>
// Public domain (Unlicense) <http://unlicense.org>
// SPDX-FileCopyrightText: 2022 Grigore Stefan <g_stefan@yahoo.com>
// SPDX-License-Identifier: Unlicense

#include <XYO/QuantumScript.hpp>
#include <XYO/QuantumScript.Extension/Console.hpp>
#include <XYO/QuantumScript.Extension/Buffer.hpp>
#include <XYO/QuantumScript.Extension/Shell.hpp>
#include <XYO/QuantumScript.Extension/ShellFind.hpp>
#include <XYO/QuantumScript.Extension/Thread.hpp>
#include <XYO/QuantumScript.Extension/Math.hpp>
#include <XYO/QuantumScript.Extension/Job.hpp>

using namespace XYO::QuantumScript;

void initExecutive(Executive *executive) {
	Extension::Console::registerInternalExtension(executive);
	Extension::Buffer::registerInternalExtension(executive);
	Extension::Shell::registerInternalExtension(executive);
	Extension::ShellFind::registerInternalExtension(executive);
	Extension::Thread::registerInternalExtension(executive);
	Extension::Math::registerInternalExtension(executive);
	Extension::Job::registerInternalExtension(executive);
};

void test(int cmdN, char *cmdS[]) {

	const char *codeFile="../../test/test.01.js";	

	if (ExecutiveX::initExecutive(cmdN, cmdS, initExecutive)) {
		ExecutiveX::includePath(Shell::getFilePath(codeFile));		
		if (ExecutiveX::executeFile(codeFile)) {
			ExecutiveX::endProcessing();
			return;
		};		
		printf("%s\n", (ExecutiveX::getError()).value());
		printf("%s", (ExecutiveX::getStackTrace()).value());			
		ExecutiveX::endProcessing();

		throw std::runtime_error("Code");
	};
};

int main(int cmdN, char *cmdS[]) {
	try {

		test(cmdN,cmdS);

		return 0;

	} catch (const std::exception &e) {
		printf("* Error: %s\n", e.what());
	} catch (...) {
		printf("* Error: Unknown\n");
	};

	return 1;
};

File: test/test.01.js
// Created by Grigore Stefan <g_stefan@yahoo.com>
// Public domain (Unlicense) <http://unlicense.org>
// SPDX-FileCopyrightText: 2016-2023 Grigore Stefan <g_stefan@yahoo.com>
// SPDX-License-Identifier: Unlicense

Script.requireExtension("Application");
Script.requireExtension("Console");
Script.requireExtension("Thread");
Script.requireExtension("Job");

var job = new Job();

job.onStart = function(process) {
	Console.writeLn("START: " + process.info);
};

job.onEnd = function(process) {
	Console.writeLn("END: " + process.info);
};

job.onTimedout = function(process) {
	Console.writeLn("TIMEDOUT: " + process.info);
};

job.addProcess("quantum-script ../../test/test.01.sub.10.js", "10");
job.addProcess("quantum-script ../../test/test.01.sub.20.js", "20");
job.addProcess("quantum-script ../../test/test.01.sub.10.js", "30-sync", "30");
job.addProcess("quantum-script ../../test/test.01.sub.10.js", "31-sync", "30");
job.addProcess("quantum-script ../../test/test.01.sub.10.js", "40-sync", "40");
job.addProcess("quantum-script ../../test/test.01.sub.10.js", "41-sync", "40");

job.addThread(function() {
	Script.requireExtension("Thread");
	CurrentThread.sleep(10 * 100);
}, null, null, "50");

job.addThread(function() {
	Script.requireExtension("Thread");
	CurrentThread.sleep(20 * 100);
}, null, null, "60");

job.addThread(function() {
	Script.requireExtension("Thread");
	CurrentThread.sleep(10 * 100);
}, null, null, "50-sync", "50");

job.addThread(function() {
	Script.requireExtension("Thread");
	CurrentThread.sleep(10 * 100);
}, null, null, "51-sync", "50");

job.process();


Console.writeLn("-> test 01 ok");
